src/main/scala/chisel3/simulator/scalatest/HasCliOptions.scala

```diff
@@ -0,0 +1,166 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.simulator.scalatest
+
+import chisel3.testing.scalatest.HasConfigMap
+import firrtl.options.StageUtils.dramaticMessage
+import org.scalatest.TestSuite
+import scala.collection.mutable
+import scala.util.control.NoStackTrace
+import svsim.Backend.HarnessCompilationFlags.enableVcdTracingSupport
+import svsim.CommonCompilationSettings.VerilogPreprocessorDefine
+import svsim.{Backend, CommonCompilationSettings}
+
+object HasCliArguments {
+
+  /** A ScalaTest command line option of the form `-D<name>=<value>`.
+    *
+    * @param name the name of the option
+    * @param convert conver the `<value>` to the internal type `A`
+    * @param updateCommonSettings a function to update the common compilation
+    * settings
+    * @param updateBackendSettings a function to update the backend-specific
+    * compilation settings
+    * @tparam the internal type of the option.  This is what the `<value>` will
+    * be converted to.
+    */
+  case class CliOption[A](
+    name:                  String,
+    help:                  String,
+    convert:               (String) => A,
+    updateCommonSettings:  (A, CommonCompilationSettings) => CommonCompilationSettings,
+    updateBackendSettings: (A, Backend.Settings) => Backend.Settings
+  )
+
+}
+
+trait HasCliArguments extends HasConfigMap { this: TestSuite =>
+
+  import HasCliArguments._
+
+  private val options = mutable.HashMap.empty[String, CliOption[_]]
+
+  final def addOption(option: CliOption[_]): Unit = {
+    if (options.contains(option.name))
+      throw new Exception("unable to add option with name '$name' because this is already taken by another option")
+
+    options += option.name -> option
+  }
+
+  private def helpBody = {
+    val optionsHelp = options.map { case (_, option) =>
+      s"""|  ${option.name}
+          |      ${option.help}
+          |""".stripMargin
+    }.mkString
+    s"""|Usage: <ScalaTest> [-D<name>=<value>...]
+        |
+        |This ChiselSim ScalaTest test supports passing command line arguments via
+        |ScalaTest's "config map" feature.  To access this, append `-D<name>=<value>` for
+        |a legal option listed below.
+        |
+        |Options:
+        |
+        |$optionsHelp""".stripMargin
+  }
+
+  private def illegalOptionCheck(): Unit = {
+    configMap.keys.foreach { case name =>
+      if (!options.contains(name)) {
+        throw new IllegalArgumentException(
+          dramaticMessage(
+            header = Some(s"illegal ChiselSim ScalaTest option '$name'"),
+            body = helpBody
+          )
+        ) with NoStackTrace
+      }
+    }
+  }
+
+  implicit def commonSettingsModifications: svsim.CommonSettingsModifications = {
+    illegalOptionCheck()
+    options.values.foldLeft(_: CommonCompilationSettings) { case (acc, option) =>
```

Comment:
```suggestion
    (settings: CommonCompilationSettings) => options.values.foldLeft(settings) { case (acc, option) =>
```
I had trouble wrapping my head around this, would prefer more explicit function syntax

Reply:
Fixed in: https://github.com/chipsalliance/chisel/pull/4777/commits/17cb9bb76863bbca2e83117bfd73c5a4ec260cdd